### PR TITLE
8271506: Add ResourceHashtable support for deleting selected entries

### PR DESCRIPTION
**Pull request by coleenp**
The ResourceHashtable doesn't have a way to delete selected entries based on things like their class has been unloaded, which is needed to replace Hashtable with ResourceHashtable.
The Nodes of the ResourceHashtable has a Key and Value of  template types.
template<typename K, typename V>
class ResourceHashtableNode : public ResourceObj {
public:
  unsigned _hash;
  K _key;
  V _value;
...
But there's no destructor so that ~K and ~V are not called (if I understand C++ correctly).

When instantiated with a value that's not a pointer, calling code does this:

  SourceObjInfo src_info(ref, read_only, follow_mode);
  bool created;
  SourceObjInfo* p = _src_obj_table.put_if_absent(src_obj, src_info, &created);

So if SourceObjInfo has a destructor, it'll have to have a careful assignment operator so that the value copied into the hashtable doesn't get deleted.

In this patch, I assign the responsibility of deleting the Key and Value of the hashtable to the do_entry function, because it's simple.  If we want to use more advanced unreadable C++ code, someone will have to suggest an alternate set of changes, because my C++ is not up to this.

Tested with tier1-3, gtest, and upcoming patch for JDK-8048190.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271506](https://bugs.openjdk.java.net/browse/JDK-8271506): Add ResourceHashtable support for deleting selected entries

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4938/head:pull/4938` \
`$ git checkout pull/4938`

Update a local copy of the PR: \
`$ git checkout pull/4938` \
`$ git pull https://git.openjdk.java.net/jdk pull/4938/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4938`

View PR using the GUI difftool: \
`$ git pr show -t 4938`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4938.diff">https://git.openjdk.java.net/jdk/pull/4938.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/resourceHash.hpp
Comment (iklam) [inline src/hotspot/share/utilities/resourceHash.hpp:219]: `unlink()` shouldn't be `const` as it modifies the table. However, you probably needed to declare it as such because it calls `bucket_at()`, which is `const`. And the reason that  `bucket_at()` is `const` is because it's called by `iterate()`, which is `const`.

The use of `const` and `const_cast` in this class is getting a bit out of whack. I have created [JDK-8271525](https://bugs.openjdk.java.net/browse/JDK-8271525) with a preliminary fix here: https://github.com/openjdk/jdk/pull/4942

Maybe we should do that before this PR? it will also make the `const_cast` unnecessary for `decrement_entries()`.

**Review by iklam — CHANGES_REQUESTED** [review]

File: src/hotspot/share/utilities/resourceHash.hpp
Comment (coleenp) [reply src/hotspot/share/utilities/resourceHash.hpp]: I agree with @tstuefe in your draft PR.  I don't think get should be non-const.  I haven't been able to parse through the monstrosity cast of lookup_node to know why and whether it should be changed.
I made unlink non-const and made the changes that he suggested there and that fixed the cast that I had to decrement the entries.

**Comment by coleenp** [comment]
I don't think this should be blocked by JDK-8271525.

**Review by iklam — APPROVED** [review]
LGTM. I agree this shouldn't be blocked by JDK-8271525.

File: src/hotspot/share/utilities/resourceHash.hpp
Comment (tstuefe) [inline src/hotspot/share/utilities/resourceHash.hpp:230]: Unless I miss it, you don't enforce the constness here, no? So, nothing prevents the called hook from modifying key or value. You may do what Ioi did in https://github.com/openjdk/jdk/pull/4942 and const-cast the parameters for the call.

**Review by tstuefe — APPROVED** [review]
Hi Coleen, 

looks good to me. One minor remark, but I leave it up to you, this is good as it is.

Thanks, Thomas

File: src/hotspot/share/utilities/resourceHash.hpp
Comment (coleenp) [reply src/hotspot/share/utilities/resourceHash.hpp:230]: I thought I answered this in email but I don't see it.  I want the do_entry function to modify the arguments, like calling their destructor or free operators.

**Comment by coleenp** [comment]
Thanks Thomas and Ioi for the code reviews.

**Comment by coleenp** [comment]
Yes, Ioi, hopefully you can just move this unlink function in your changes.
/integrate

**Pull request closed without merging**
